src/reducers/computeRenderedRows.js

```diff
@@ -216,26 +223,103 @@ function computeRenderedRowOffsets(state, rowRange) {
-        firstViewportIdx,
-        endViewportIdx - 1,
-        rowIdx
-      );
-    }
+  state.rowHeights = rowOffsetsCache;
+  state.rows = bufferMapping;
+}
+
+/**
+ * Walk the rows to render that are present in the viewport and compute the height offsets and
+ * positions in the row buffer. (Also modifies the height offsets for the rows outside the viewport, to avoid
+ * overlapping).
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * TODO (pradeep): Refactor computeViewPortRenderedRowOffsets and computeRenderRowOffsets. There's some code
+ * duplication present.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdx: number,
+ *   endViewportIdx: number,
+ *   firstBufferIdx: number,
+ *   firstViewportIdx: number,
+ * }} rowRange
+ * @private
+ */
+function computeViewPortRenderedRowOffsets(state, rowRange) {
+  const { bufferSet, rowOffsets, storedHeights } = state;
+  const {
+    endBufferIdx,
+    endViewportIdx,
+    firstBufferIdx,
+    firstViewportIdx,
+  } = rowRange;
+
+  const renderedRowsCount = endBufferIdx - firstBufferIdx;
+  if (renderedRowsCount === 0) {
+    state.rowHeights = {};
+    state.rows = [];
+    return;
+  }
+
+  // output for this function
+  const bufferMapping = state.rows;
+  const rowOffsetsCache = {};
+
+  // Used as a way to get the row ID deltas across scrolls
+  const bufferMappingOld = state.rows.slice();
+  const rowOffsetsCacheOld = Object.assign({}, state.rowHeights);
+
+  // incremental way for calculating rowOffset
+  let runningOffset = rowOffsets.sumUntil(firstBufferIdx);
 
-    // If we can't reuse any existing position, create a new one
-    if (rowPosition === null) {
-      rowPosition = bufferSet.getNewPositionForValue(rowIdx);
+  for (let rowIdx = firstBufferIdx; rowIdx < endBufferIdx; rowIdx++) {
+
+    // Update the offset for rendering the row
+    rowOffsetsCache[rowIdx] = runningOffset;
+    runningOffset += storedHeights[rowIdx];
+
+    // we'll skip rows inside the viewport
+    if (rowIdx < firstViewportIdx || rowIdx >= endViewportIdx) {
+      continue;
     }
 
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, firstViewportIdx, endViewportIdx, bufferSet, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  // Each time we make a set of row offsets, but we only change the position for the rows inside the buffer
+  // there's a possibility that the old rows (outside the viewport) overlap the new ones since scrollY was changed.
+  // To avoid this, we need to update the rowOffsets for them.
+  const unchangedRows = intersection(bufferMapping, bufferMappingOld);
+  unchangedRows.forEach(rowIdx => rowOffsetsCache[rowIdx] = rowOffsetsCacheOld[rowIdx]);
+
   state.rowHeights = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+function addRowToBuffer(rowIdx, bufferSetStartIndex, bufferSetEndIndex, bufferSet, maxBufferSize) {
```

Comment:
This function simply adds the row ID into the buffer, and returns the new index.
Used by both computeRenderedRowOffsets, and computeViewPortRenderedRowOffsets.